docs/isl-2-0/spec.md

```diff
@@ -0,0 +1,668 @@
+|----:|------------------------------|
+| `i` | case insensitive             |
+| `m` | ^ and $ match at line breaks |
+
+The following classes are provided:
+
+| | |
+| `\d` | digit: `[0-9]` |
+| `\D` | non-digit |
+| `\s` | whitespace: `[ \f\n\r\t]` |
+| `\S` | non-whitespace |
+| `\w` | word character: `[A-Za-z0-9_]` |
+| `\W` | non-word character |
+
+The following characters may be escaped with a backslash:  `. ^ $ | ? * + \ [ ] ( ) { }`.
+Note that in Ion text a backslash must itself be escaped, so correct escaping of these characters requires two backslashes, e.g.:  `\\.`.
+
+
+## timestamp_offset
+
+{% include grammar-element.md productions="timestamp_offset" %}
+
+Limits the timestamp offsets that are allowed. 
+An offset is specified as a string of the form `"[+|-]hh:mm"`, where `hh` is a two-digit number between 00 and 23, inclusive, and `mm` is a two-digit number between 00 and 59, inclusive.
+The offset `"+00:00"` is equivalent to `Z`.
+The offset `"-00:00"` represents unknown local offset.
+For more details about offsets, see the [Ion Specification – Timestamps](https://amzn.github.io/ion-docs/docs/spec.html#timestamp).
+
+{% comment %}{% include example.md title="`timestamp_offset`" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+## timestamp_precision
+
+{% include grammar-element.md productions="timestamp_precision,timestamp_precision_value,range_timestamp_precision" %}
+
+Indicates the exact or minimum/maximum precision of a timestamp.
+Valid precision values are, in order of increasing precision: `year`, `month`, `day`, `minute`, `second`, `millisecond`, `microsecond`, and `nanosecond`.
+
+{% comment %}{% include example.md title="`timestamp_precision`" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+## type
+
+{% include grammar-element.md productions="type" %}
+
+Indicates the type that a value shall be validated against.
+
+{% comment %}{% include example.md title="`type`" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+## utf8_byte_length
+
+{% include grammar-element.md productions="utf8_byte_length" %}
+
+An exact or minimum/maximum indicating number of bytes in the [UTF-8](https://en.wikipedia.org/wiki/UTF-8) representation of a string or symbol.
+The values `null.string` and `null.symbol` do not have a length of 0.
+Rather, they have no length at all, and are always invalid for this constraint.
+
+{% comment %}{% include example.md title="`utf8_byte_length`" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+## valid_values
+
+{% include grammar-element.md productions="valid_values,value_or_range" %}
+
+A list of acceptable, non-annotated values;  any values not present in the list are invalid.
+The argument to this constraint can be a range or a list of unannotated values or a list containing a mix of ranges and unannotated values.
+
+Ignoring annotations, the value must match one of the list of valid values or be within the boundaries of a range.
```

Comment:
The range included in the `valid_values` does not follow the behaviors of constraint `range` right?

Reply:
I'm not sure I understand what it is that you are asking. Do you mind clarifying, maybe with an example if possible?

Reply:
> The range included in the `valid_values` does not follow the behaviors of constraint `range` right?

Sorry for the unclear question. The argument of `valid_values` could be a `range`: i.e. `valid_values: range::[0, 3] `which indicates the valid value should be within this range, and my understanding is the end values of this range  should be exclusive by default. According to the `Ranges` section, if we desire to exclude the end values, we should annotate the value with 'exclusive' and the default behavior is for both ends of the range to be 'inclusive'. 

TL;DR: What is the default behavior of the `range` included in the `valid_values`? Is that aligned with the `Ranges` we defined before?

Reply:
Yes, it is aligned. All ranges have inclusive endpoints unless annotated by `exclusive::`.